Osfv ci tests scope determining

philipanda opened the pull request:
<!--
SPDX-FileCopyrightText: 2024 3mdeb <contact@3mdeb.com>

SPDX-License-Identifier: Apache-2.0
-->

<!-- markdownlint-disable MD041 -->
## Short description
The script is supposed to help determine which tests should be run in order to verify if the changes performed to OSFV are breaking anything.
The script's functionality is tested using `scripts/ci/regression-scope/regression_scope_selftests.py`, which also serves a purpose of demonstrating how the rules files can be created.
## Related Issue(s)

philipanda committed:
regression-scope/rules.json: Add example rules file

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

rules.json: simplify

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

ci/regression-scope/rules.json: Add specifying dirs for deps search

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

regression-scope/rules.json: Change key 'command' to 'custom_command'

To emphasize more how it overrides the default behavior of running robot
framework

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

regression-scope/rules.json: Include all the test suites

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
osfv-regression-scope.py: Add prototype script for parsing rules

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

osfv-regression-scope.py: Make a CLI out of it

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

osfv-regression-scope.py: Add option to compare against any revision

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

osfv-regression-scope.py: Fix misleading synopsis

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

osfv-regression-scope.py: Add checking for deps and simple self-tests

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

osfv_regression_scope.py: Fix help messages

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

osfv_regression_scope.py: Change output to simple strings

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
requirements.txt: Add fire

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
ci/regression-scope: Allow adding robot args to rules

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

scripts/ci/regression-scope: Revert platform-configs tests

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
scripts/ci/regression-scope: Add printing args for robot wrapper run.sh

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
scripts/ci/regression-scope: Allow multpiple "runs" per rule

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

scripts/ci/regression-scope: Rename file match modes to more meaningful

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

scripts/ci/regression-scope/: Remove unused functionality

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
.github/PULL_REQUEST_TEMPLATE: Add PR templates

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

filipleple commented:
This looks great, I'm just not sure if there isn't a chance that a single test file gets picked multiple times for the same PR? As in, if a large PR touches multiple files that are explicitly matched by one rule and implicitly used (via FILE_CONTAINS_MATCH) by another.

Then we’d get repeated execution of the same test suite, increasing CI time without benefit.

philipanda commented:
>This looks great, I'm just not sure if there isn't a chance that a single test file gets picked multiple times for the same PR? As in, if a large PR touches multiple files that are explicitly matched by one rule and implicitly used (via FILE_CONTAINS_MATCH) by another.
>
>Then we’d get repeated execution of the same test suite, increasing CI time without benefit.

Creating two rules that both suggest running one test suite **WILL** result in the same suite appearing two times in the output of either `filenames`, `commands`, or `robot_wrapper_args`.

Eg. after modifying `dasharo-performance/boot-time-measure.robot` and `lib/cbmem.robot`:
```
$ python scripts/ci/regression-scope/osfv_regression_scope.py commands
export RTE_IP=127.0.0.1 export FW_FILE=scripts/ci/qemu_q35.rom export CONFIG=qemu
scripts/run.sh dasharo-performance/boot-time-measure.robot -- -v snipeit:no
export RTE_IP=127.0.0.1 export FW_FILE=scripts/ci/qemu_q35.rom export CONFIG=qemu
scripts/run.sh dasharo-performance/boot-time-measure.robot dasharo-performance/fast-boot.robot -- -v snipeit:no
$
```

Two distinct rules can define that the suites matched by them should be run with arbitrary optional robot arguments and with arbitrary environment variables set.

For this reason I am not sure if preventing this could be done in a way smarter than just looking for duplicated sets of {ENV_VARS+ROBOT_ARGS+SUITE} in the sum of the outputs of all the parsed rules.

Nevertheless that is something worth optimizing.

philipanda committed:
ci/regression-scope/lib/parser_manager.py: Add library

It manages multiple rule parsers and makes sure one test suite is not
run multiple times with the same conditions

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
scripts/ci/regression-scope/: Adjust to use the lib/parser_manager.py

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

regression_scope_selftestsppy: fix duplicated test function from rebase

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda commented:
@filipleple In the last few commits I've added what you've asked for.
If multiple rules would result in running a test case with the exact same environment variables and robot arguments, then it will not be executed more than once. If the environment variables or robot arguments differ, every unique requested combination will be run. There are two test cases to cover this https://github.com/Dasharo/open-source-firmware-validation/pull/841/files#diff-a92cab7ff5d4000dfd89849ecd23ee450590d996f0026ae4b38fc9fbb2d75e7fR342-R473

filipleple reviewed (approved):
Neat, LGTM now :fire:

macpijan reviewed (commented):
Changes look good, but maybe we should squash some of these commits prior merging?

philipanda commented:
@macpijan I've just squashed the commits, is that enough?
